proposals/p2200.md

```diff
@@ -0,0 +1,1117 @@
+# Template generics
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/2200)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Abstract](#abstract)
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Terminology](#terminology)
+-   [Proposal](#proposal)
+-   [Details](#details)
+    -   [Syntax](#syntax)
+        -   [Branching on type identity](#branching-on-type-identity)
+    -   [Value phases](#value-phases)
+    -   [`auto`](#auto)
+    -   [Template constraints](#template-constraints)
+    -   [Name lookup](#name-lookup)
+    -   [Transition from C++ templates to Carbon checked generics](#transition-from-c-templates-to-carbon-checked-generics)
+        -   [To template Carbon with structural constraints](#to-template-carbon-with-structural-constraints)
+        -   [To interface constraints](#to-interface-constraints)
+        -   [To checked generic](#to-checked-generic)
+    -   [Validity can depend on value](#validity-can-depend-on-value)
+    -   [Template dependent](#template-dependent)
+        -   [Simple member access](#simple-member-access)
+        -   [Compound member access](#compound-member-access)
+        -   [Impl lookup](#impl-lookup)
+        -   [Use of dependent value is dependent](#use-of-dependent-value-is-dependent)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+    -   [Only checked generics](#only-checked-generics)
+    -   [SFINAE](#sfinae)
+    -   [Ad hoc API specialization](#ad-hoc-api-specialization)
+    -   [Value phase of bindings determined by initializer](#value-phase-of-bindings-determined-by-initializer)
+    -   [Simpler template dependent rules that delayed more checking](#simpler-template-dependent-rules-that-delayed-more-checking)
+-   [Future work](#future-work)
+    -   [Expanded template constraints](#expanded-template-constraints)
+    -   [Predicates: constraints on values](#predicates-constraints-on-values)
+    -   [Checked generics calling templates](#checked-generics-calling-templates)
+    -   [Which expressions will be evaluated at compile time](#which-expressions-will-be-evaluated-at-compile-time)
+
+<!-- tocstop -->
+
+## Abstract
+
+Add template generics, with optional constraints but no SFINAE, to Carbon.
```

Comment:
```suggestion
Add template generics, with optional constraints but no [SFINAE](https://en.cppreference.com/w/cpp/language/sfinae), to Carbon.
```

(Or a forward link into the terminology section which I see has a great definition here, mostly just trying to solve the forward-reference.)